### PR TITLE
feat: regex sampler

### LINKED ISSUE #152: Generated examples are not valid when using regular expression patterns
Generated examples are not valid when using regular expression patterns.
In this case, the example is simply "string" without honoring the regular expression.

Steps to replicate:
1. OA as an example:
```
openapi: 3.0.3
info:
  version: 1.0.0
  title: Test
  description: Test
servers:
  - url: "/api/1"
paths:
  /test:
    get:
      operationId: Test
      summary: Test
      description: Test
      responses:
        "200":
          description: Test
          content:
            application/json:
              schema:
                type: object
                properties:
                  test:
                    type: string
                    pattern: "^[0-9]{5,8}$"
```
Observe the example rendered as `{"test": "string"}` that is not valid against the schema.

image

Original issue: https://github.com/Redocly/redoc/issues/2312

### PR DESCRIPTION
**Pull request by llllvvuu**
Vendors a tiny dependency (total dependency tree <10 KB
uncompressed).

Other than removing unused code from the dependency, the only change
made is to change the following line:

https://github.com/fent/randexp.js/blob/2b35ea607883fa8cafa63bd3bdb5c12d695f873a/lib/randexp.js#L109

to a separately customizable callback, in order to adjust the length to the minLength and maxLength constraints.

## What/Why/How?
Fixes #152 and https://github.com/Redocly/redoc/issues/2312

## Reference
#152

## Testing
Added unit tests and fuzz tests for both basic and advanced cases.

## Screenshots (optional)
N/A

## Check yourself

- [x] Code is linted
- [x] Tested
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
Regex denial-of-service applies as usual, ensure code is not running on untrusted input in a place where a hanging process would be harmful.
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -88,6 +88,7 @@
   },
   "dependencies": {
     "@types/json-schema": "^7.0.7",
-    "json-pointer": "0.6.2"
+    "json-pointer": "0.6.2",
+    "randexp": "^0.5.3"
```
Comment (RomanHotsiy) [inline package.json:92]: This is strange to include this as dependency but not use it directly. I think it's better to add 2 other dependencies here.

But I'm also concerned by those dependencies. They are not very active. Last update 5 years ago.

**Review by RomanHotsiy — COMMENTED** [review]
This looks more or less good but I'm not sure if the complexity involved is worth it.
People can always provide example manually which will be better quality.

**Comment by llllvvuu** [comment]
That's a valid concern.

One way to reduce the complexity would be to simply ignore the min_length and max_length for regexes. Another way could be to copy faker.js fromRegExp, which is less complete but has no dependencies:

https://github.com/faker-js/faker/blob/fd6557e4b3716ebe41aa38a58b4ae112e6baf0d2/src/modules/helpers/index.ts#L401-L580

**Comment by RomanHotsiy** [comment]
I would cope the faker-js one. This looks good.

**Comment by llllvvuu** [comment]
Done! It's definitely less capable than randexp but I have added to the README a link to their docs which note the limitations.

File: src/vendor/fakerFromRegexp.js [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/** @returns {boolean} */
+function randBool() {
```
Comment (RomanHotsiy) [inline src/vendor/fakerFromRegexp.js:66]: Oh, so this tool is not deterministic 😢 

Can we adjust it to produce same result always.

One of the features of this tools is "deterministic" (check the readme, it's first on the list).

There are multiple reasons to want determinism in such tool.

File: src/samplers/string.js [hunk 3]
```diff
@@ -1,9 +1,18 @@
 'use strict';
 
 import { ensureMinLength, toRFCDateTime, uuid } from '../utils';
+import { fromRegExp } from '../vendor/fakerFromRegexp';
```
Comment (RomanHotsiy) [inline src/samplers/string.js:4]: Let's make it a bit more consistent.

let's move it from vendor folder into `samplers/string-regex.js` and rename the internal input from `fromRegExp` to `regexSample`.

File: src/samplers/string.js [hunk 4]
```diff
@@ -96,8 +101,8 @@ function relativeJsonPointerSample() {
   return '1/relative/json/pointer';
 }
 
-function regexSample() {
-  return '/regex/';
+function regexSample(min, max, _propertyName, pattern) {
+  return pattern ? fromRegExp(pattern) : truncateString('/regex/', min, max)
```
Comment (RomanHotsiy) [inline src/samplers/string.js:105]: we don't need to adjust it. Let's leave it is.
This is a string with `format: regex` which means it has to be a valid regex. I doubt someone will provide regex for regex.

**Review by RomanHotsiy — CHANGES_REQUESTED** [review]

File: src/vendor/fakerFromRegexp.js (hunk 2 above)
Comment (llllvvuu) [reply src/vendor/fakerFromRegexp.js]: Yes, adjusted

File: src/samplers/string.js (hunk 3 above)
Comment (llllvvuu) [reply src/samplers/string.js]: OK, adjusted. I could not import as just `regexSample` due to naming conflict so I did a namespace import

File: src/samplers/string.js (hunk 4 above)
Comment (llllvvuu) [reply src/samplers/string.js]: Ah, I misunderstood what `format: regex` meant. Thank you, fixed

**Review by RomanHotsiy — APPROVED** [review]

**Comment by RomanHotsiy** [comment]
@Redocly/keyboard-warriors please, merge it when you're ready 🙏

**Comment by kanvik175** [comment]
we really need this feature 😢. when can you merge it? 🙏

**Review by AlexVarchuk — APPROVED** [review]

**Comment by lornajane** [comment]
Thanks @llllvvuu for adding this feature!

**Pull request merged**